src/main/java/net/logstash/logback/pattern/AbstractJsonPatternParser.java

```diff
@@ -259,323 +170,380 @@ protected Object transform(final String value) throws IOException {
     }
-
-    private ChildrenWriter parseChildren(JsonNode node) {
-        List<FieldWriter<Event>> children = new ArrayList<>();
-        for (Iterator<Map.Entry<String, JsonNode>> nodeFields = node.fields(); nodeFields.hasNext();) {
-            Map.Entry<String, JsonNode> field = nodeFields.next();
-
-            String key = field.getKey();
-            JsonNode value = field.getValue();
-
-            if (value.isTextual()) {
-                ValueGetter<?, Event> getter = makeComputableValueGetter(value.asText());
-                children.add(new ComputableObjectFieldWriter(key, getter));
-            } else {
-                children.add(new DelegatingObjectFieldWriter(key, parseValue(value)));
-            }
+    
+    
+    private static class DelegatingNodeWriter<Event> implements NodeWriter<Event> {
+        private final NodeWriter<Event> delegate;
+        
+        DelegatingNodeWriter(NodeWriter<Event> delegate) {
+            this.delegate = Objects.requireNonNull(delegate);
+        }
+        
+        @Override
+        public void write(JsonGenerator generator, Event event) throws IOException {
+            this.delegate.write(generator, event);
         }
-        return new ChildrenWriter(children);
     }
-
     
-    public NodeWriter<Event> parse(String pattern) {
-        if (StringUtils.isEmpty(pattern)) {
-            return null;
+    
+    private static class OmitEmptyFieldWriter<Event> extends DelegatingNodeWriter<Event> {
+        private static final ThreadLocal<ReusableFilteringGenerator> filteringGenerators = ThreadLocal.withInitial(ReusableFilteringGenerator::new);
+        
+        OmitEmptyFieldWriter(NodeWriter<Event> delegate) {
+            super(delegate);
         }
-
-        final ObjectNode node;
-        try (JsonParser jsonParser = jsonFactory.createParser(pattern)) {
-            node = JsonReadingUtils.readFullyAsObjectNode(jsonFactory, pattern);
-        } catch (IOException e) {
-            contextAware.addError("[pattern] is not a valid JSON object", e);
-            return null;
+        
+        @Override
+        public void write(JsonGenerator generator, Event event) throws IOException {
+            ReusableFilteringGenerator filteringGenerator = filteringGenerators.get();
+            try {
+                filteringGenerator.connect(generator);
+                super.write(filteringGenerator, event);
+                
+            } catch (RuntimeException | IOException e) {
+                filteringGenerators.remove();
+                throw e;
+                
+            } finally {
+                filteringGenerator.disconnect();
+            }
         }
-
-        return parseChildren(node);
     }
-
-    /**
-     * Return true if the given value is considered to be "empty".
-     * 
-     * @param value value to inspect
-     * @return {@code true} if the given value is considered to be "empty".
-     */
-    private boolean isEmptyValue(Object value) {
```

Comment:
The isEmptyValue method would result in filtering out empty collections, maps, and container nodes.  Does the new code still do that?  (I couldn't easily see where they were filtered out.  I only saw filtering out of null and empty strings)

Reply:
Yes it does... otherwise the following test would fail: https://github.com/logstash/logstash-logback-encoder/blob/patternlayout-refactoring/src/test/java/net/logstash/logback/pattern/AbstractJsonPatternParserTest.java#L345 

> I couldn't easily see where they were filtered out.
Me neither... I did not check how the filtering mechanism works in detail but I notice empty collections were filtered out ;-)